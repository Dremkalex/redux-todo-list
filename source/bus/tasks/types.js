export const types = {
    // Sync
    FILL_TASKS:               'FILL_TASKS',
    CREATE_TASK:              'CREATE_TASK',
    REMOVE_TASK:              'REMOVE_TASK',
    UPDATE_TASK:              'UPDATE_TASK',
    COMPLETE_ALL_TASK:        'COMPLETE_ALL_TASK',
    UPDATE_NEW_TASK_MESSAGE:  'UPDATE_NEW_TASK_MESSAGE',
    UPDATE_EDIT_TASK_MESSAGE: 'UPDATE_EDIT_TASK_MESSAGE',
    UPDATE_TASKS_FILTER:      'UPDATE_TASKS_FILTER',
    SET_EDIT_TASK_ID:         'SET_EDIT_TASK_ID',
    RESET_EDIT_TASK_ID:       'RESET_EDIT_TASK_ID',

    // Async
    FETCH_TASKS_ASYNC:       'FETCH_TASKS_ASYNC',
    CREATE_TASK_ASYNC:       'CREATE_TASK_ASYNC',
    REMOVE_TASK_ASYNC:       'REMOVE_TASK_ASYNC',
    UPDATE_TASK_ASYNC:       'UPDATE_TASK_ASYNC',
    COMPLETE_ALL_TASK_ASYNC: 'COMPLETE_ALL_TASK_ASYNC',
};
