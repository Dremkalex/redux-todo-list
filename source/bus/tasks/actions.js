// Types
import { types } from './types';

export const tasksActions = {
    // Sync
    fillTasks: (tasks) => ({
        type:    types.FILL_TASKS,
        payload: tasks,
    }),

    createTask: (task) => ({
        type:    types.CREATE_TASK,
        payload: task,
    }),

    removeTask: (taskId) => ({
        type:    types.REMOVE_TASK,
        payload: taskId,
    }),

    updateTask: (updatedTask) => ({
        type:    types.UPDATE_TASK,
        payload: updatedTask,
    }),

    updateNewTaskMessage: (message) => ({
        type:    types.UPDATE_NEW_TASK_MESSAGE,
        payload: message,
    }),

    updateEditTaskMessage: (message) => ({
        type:    types.UPDATE_EDIT_TASK_MESSAGE,
        payload: message,
    }),

    updateTasksFilter: (message) => ({
        type:    types.UPDATE_TASKS_FILTER,
        payload: message,
    }),

    setEditTaskId: (editTaskId) => ({
        type:    types.SET_EDIT_TASK_ID,
        payload: editTaskId,
    }),

    resetEditTaskId: () => ({
        type: types.RESET_EDIT_TASK_ID,
    }),

    // Async
    fetchTasksAsync: () => ({
        type: types.FETCH_TASKS_ASYNC,
    }),

    createTaskAsync: (newTaskMessage) => ({
        type:    types.CREATE_TASK_ASYNC,
        payload: newTaskMessage,
    }),

    removeTaskAsync: (taskId) => ({
        type:    types.REMOVE_TASK_ASYNC,
        payload: taskId,
    }),

    updateTaskAsync: (updatedTask) => ({
        type:    types.UPDATE_TASK_ASYNC,
        payload: updatedTask,
    }),

    completeAllTasksAsync: (tasks) => ({
        type:    types.COMPLETE_ALL_TASK_ASYNC,
        payload: tasks,
    }),
};
