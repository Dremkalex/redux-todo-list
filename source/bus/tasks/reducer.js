// Core
import { fromJS, Map, List } from 'immutable';

// Types
import { types } from './types';

const initialState = Map({
    tasks:           List(),
    newTaskMessage:  '',
    editTaskMessage: '',
    editTaskId:      '',
    tasksFilter:     '',
});

export const tasksReducer = (state = initialState, { type, payload }) => {
    switch (type) {
        case types.FILL_TASKS:
            return state.set('tasks', fromJS(payload));

        case types.CREATE_TASK:
            return state.set('tasks', state.get('tasks').unshift(fromJS(payload)));

        case types.REMOVE_TASK:
            return state.set('tasks',
                state.get('tasks').
                    filter((task) => task.get('id')!== payload));

        case types.UPDATE_TASK:
            return state.set('tasks',
                state.get('tasks')
                    .update(state.get('tasks').findIndex((task) => task.get('id') === payload.id),
                        (task) => task.merge(payload)));

        case types.UPDATE_NEW_TASK_MESSAGE:
            return state.set('newTaskMessage', payload);

        case types.UPDATE_EDIT_TASK_MESSAGE:
            return state.set('editTaskMessage', payload);

        case types.SET_EDIT_TASK_ID:
            return state.set('editTaskId', payload);

        case types.RESET_EDIT_TASK_ID:
            return state.set('editTaskId', '');

        case types.UPDATE_TASKS_FILTER:
            return state.set('tasksFilter', payload);

        default:
            return state;
    }
};
