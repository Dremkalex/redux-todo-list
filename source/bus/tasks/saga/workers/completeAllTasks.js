// Core
import { put, apply } from 'redux-saga/effects';

// Instruments
import { api } from '../../../../REST';

// Actions
import { uiActions } from '../../../ui/actions';
import { tasksActions } from '../../actions';

export function* completeAllTasks ({ payload: tasks }) {
    yield put(uiActions.startFetching());

    try {
        const completedTasks = yield tasks.map((task) => task.set('completed', true));
        const response = yield apply(api, api.tasks.completeAllTasks, [completedTasks]);
        const { data: completedTasksFromResponse, message } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(message);
        }

        yield put(tasksActions.fillTasks(completedTasksFromResponse));
    } catch (error) {
        yield put(uiActions.emitError(error, 'completeAllTasks worker'));
    } finally {
        yield put(uiActions.stopFetching());
    }
}
