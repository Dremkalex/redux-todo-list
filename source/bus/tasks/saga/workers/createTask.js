// Core
import { put, apply } from 'redux-saga/effects';

// Instruments
import { api } from '../../../../REST';

// Actions
import { uiActions } from '../../../ui/actions';
import { tasksActions } from '../../../tasks/actions';

export function* createTask ({ payload: newTaskMessage }) {
    yield put(uiActions.startFetching());

    try {
        const response = yield apply(api, api.tasks.create, [newTaskMessage]);
        const { data: task, message } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(message);
        }

        yield put(tasksActions.createTask(task));
        yield put(tasksActions.updateNewTaskMessage(''));
    } catch (error) {
        yield put(uiActions.emitError(error, 'createTask worker'));
    } finally {
        yield put(uiActions.stopFetching());
    }
}
