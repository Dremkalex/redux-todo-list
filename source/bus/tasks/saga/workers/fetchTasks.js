// Core
import { put, apply } from 'redux-saga/effects';

// Instruments
import { api } from '../../../../REST';

// Actions
import { uiActions } from '../../../ui/actions';
import { tasksActions } from '../../../tasks/actions';

export function* fetchTasks () {
    yield put(uiActions.startFetching());

    try {
        const response = yield apply(api, api.tasks.fetch);
        const { data: tasks, message } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(message);
        }

        yield put(tasksActions.fillTasks(tasks));

    } catch (error) {
        yield put(uiActions.emitError(error, 'fetchTasks worker'));

    } finally {
        yield put(uiActions.stopFetching());
    }
}
