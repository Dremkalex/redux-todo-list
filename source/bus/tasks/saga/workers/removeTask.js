// Core
import { put, apply } from 'redux-saga/effects';

// Instruments
import { api } from '../../../../REST';

// Actions
import { uiActions } from '../../../ui/actions';
import { tasksActions } from '../../../tasks/actions';

export function* removeTask ({ payload: taskId }) {
    yield put(uiActions.startFetching());

    try {
        const response = yield apply(api, api.tasks.remove, [taskId]);

        if (response.status !== 204) {
            const { message } = yield apply(response, response.json);

            throw new Error(message);
        }

        yield put(tasksActions.removeTask(taskId));
    } catch (error) {
        yield put(uiActions.emitError(error, 'removeTask worker'));
    } finally {
        yield put(uiActions.stopFetching());
    }
}
