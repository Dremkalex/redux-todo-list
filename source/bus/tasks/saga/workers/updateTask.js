// Core
import { put, apply } from 'redux-saga/effects';

// Instruments
import { api } from '../../../../REST';

// Actions
import { uiActions } from '../../../ui/actions';
import { tasksActions } from '../../../tasks/actions';

export function* updateTask ({ payload: updatedTask }) {
    yield put(uiActions.startFetching());

    try {
        const response = yield apply(api, api.tasks.update, [updatedTask]);
        const { data: [updatedTaskFromResponse], message } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(message);
        }

        yield put(tasksActions.updateTask(updatedTaskFromResponse));
    } catch (error) {
        yield put(uiActions.emitError(error, 'updateTask worker'));
    } finally {
        yield put(tasksActions.resetEditTaskId());
        yield put(uiActions.stopFetching());
    }
}
