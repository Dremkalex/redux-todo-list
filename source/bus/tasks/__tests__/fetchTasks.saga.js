// Core
import { put, apply } from "redux-saga/effects";
import { cloneableGenerator } from "redux-saga/utils";

// Instruments
import { api } from "../../../REST";
import { uiActions } from "../../ui/actions";
import { tasksActions } from "../../tasks/actions";
import { fetchTasks } from "../saga/workers";

const fetchTasksAction = tasksActions.fetchTasksAsync();
const saga = cloneableGenerator(fetchTasks)(fetchTasksAction);
let clone = null;

describe("fetchTasks saga:", () => {
    describe("should pass until response received", () => {
        test('should dispatch "startFetching" action', () => {
            expect(saga.next().value).toEqual(put(uiActions.startFetching()));
        });

        test("should call a fetch request", () => {
            expect(saga.next().value).toEqual(apply(api, api.tasks.fetch));
            clone = saga.clone();
        });
    });

    describe("should handle a 400 status response", () => {
        test("the fetch request should return 400 status response", () => {
            expect(clone.next(__.fetchResponseFail400).value).toEqual(
                apply(__.fetchResponseFail400, __.fetchResponseFail400.json)
            );
        });

        test("should contain a response data object", () => {
            expect(clone.next(__.responseDataFail).value).toEqual(
                put(uiActions.emitError(__.error, 'fetchTasks worker'))
            );
        });

        test('should dispatch "stopFetching" action', () => {
            expect(clone.next().value).toEqual(put(uiActions.stopFetching()));
        });

        test("should finish", () => {
            expect(clone.next().done).toBe(true);
        });
    });

    describe("should handle a 200 status response", () => {
        test("the fetch request should return 200 status response", () => {
            expect(saga.next(__.fetchResponseSuccess).value).toEqual(
                apply(__.fetchResponseSuccess, __.fetchResponseSuccess.json)
            );
        });

        test('should dispatch "fillTasks" action', () => {
            expect(saga.next(__.responseDataSuccess).value)
                .toMatchSnapshot();
        });

        test('should dispatch "stopFetching" action', () => {
            expect(saga.next().value).toMatchSnapshot();
        });

        test("should finish", () => {
            expect(clone.next().done).toBe(true);
        });
    });
});
