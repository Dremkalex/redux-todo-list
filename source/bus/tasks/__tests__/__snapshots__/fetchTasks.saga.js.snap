// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTasks saga: should handle a 200 status response should dispatch "fillTasks" action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "error": true,
      "meta": "fetchTasks worker",
      "payload": [TypeError: Cannot destructure property \`data\` of 'undefined' or 'null'.],
      "type": "EMIT_ERROR",
    },
    "channel": null,
  },
}
`;

exports[`fetchTasks saga: should handle a 200 status response should dispatch "stopFetching" action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "STOP_FETCHING",
    },
    "channel": null,
  },
}
`;
